hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/MemberSchemaService.java

```diff
@@ -131,47 +95,194 @@ public void put(Schema schema) {
+        }
+
+        return new SendSchemaReplicationsOperation(replications);
+    }
+
+    /**
+     * Called in the joining member to replay all the replications that were
+     * available in the master node, when the pre-join operation is prepared.
+     * <p>
+     * It puts the schemas to the in-memory registry of the joining member,
+     * persists them to HotRestart, and updates the local replicator with the
+     * replications using the same status as they are sent from the master.
+     * <p>
+     * See the documentation of {@link SendSchemaReplicationsOperation} to see
+     * the idea behind this.
+     *
+     * @param replications to replay.
+     */
+    public void replayReplications(Collection<SchemaReplication> replications) {
+        for (SchemaReplication replication : replications) {
+            Schema schema = replication.getSchema();
+            putLocal(schema);
+        }
+        List<Schema> schemas = replications.stream()
+                .map(SchemaReplication::getSchema)
+                .collect(Collectors.toList());
+
+        persistAllSchemasToHotRestart(schemas);
+        replicator.setReplications(replications);
+    }
+
+    /**
+     * Called on the participant members to prepare for the replication process
+     * by:
+     * <ul>
+     *     <li>Putting the schema to an in-memory registry</li>
+     *     <li>Persisting the schema to HotRestart(if available)</li>
+     *     <li>Marking the schema as {@link SchemaReplicationStatus#PREPARED}
+     *     </li>
+     * </ul>
+     *
+     * @param schema to be prepared for replication.
+     */
+    public void onSchemaPreparationRequest(Schema schema) {
+        // If it is already PREPARED or REPLICATED, do nothing
+        if (replicator.getReplicationStatus(schema) != null) {
+            return;
+        }
+
+        putLocal(schema);
+        persistSchemaToHotRestart(schema);
+        replicator.markSchemaAsPrepared(schema);
+    }
+
+    /**
+     * Called on the participant members to mark the replication status of the
+     * schema as {@link SchemaReplicationStatus#REPLICATED}.
+     *
+     * @param schemaId                  of the schema that is replicated to the
+     *                                  cluster.
+     * @param isLocalOperationExecution {@code true} if the call is made through
+     *                                  local execution of the operation
+     */
+    public void onSchemaAckRequest(long schemaId, boolean isLocalOperationExecution) {
+        replicator.markSchemaAsReplicated(schemaId, isLocalOperationExecution);
+    }
+
+    /**
+     * Called when the schemas are read from the HotRestart data.
+     * <p>
+     * Each schema will be put into the in-memory registry and will be marked as
+     * {@link SchemaReplicationStatus#PREPARED}, as we cannot distinguish
+     * schemas that are {@link SchemaReplicationStatus#PREPARED} and
+     * {@link SchemaReplicationStatus#REPLICATED} from the HotRestart data.
```

Comment:
why don't we keep the state in the hotrestart file or why don't we only write the schemas with status `REPLICATED` to hotrestart file?

Reply:
Because we write to HotRestart in the preparation phase.